### ISSUE TITLE
[Dynamic Fuzzing] Universal PyTorch Forward Hook Instrumentation

### ISSUE THREAD

**Issue by KrishG7** [KrishG7/Aegis-Tensor#4]
### Assignee: @7dxksh7
**Subsystem:** Dynamic Fuzzing & Activation Hook Engine  
**Sprint Due Date:** September 22, 2026 (Milestone 1)

---

### 1. Problem & Architecture Context
Aegis-Tensor monitors intermediate activations in real time using PyTorch `register_forward_hook`. Because modern architectures (Vision Transformers, BERT, LLMs) return complex non-standard output structures (tuples, HuggingFace `ModelOutput` objects), the hook manager must safely unpack tensor representations without crashing or leaking GPU memory.

---

### 2. Implementation Steps
1. **Enhance `ActivationHookManager` in `aegis/fuzzer.py`:**
   - Unpack complex outputs: `tuple`, `list`, dataclasses (`output.last_hidden_state`), and nested dictionaries.
   - Detach tensors immediately: `output.detach().float().cpu()` to prevent autograd graph construction and RAM/VRAM exhaustion.

2. **Hook Filtering & Module Selection:**
   - Add filtering option to hook specific module types (e.g. `nn.Linear`, `nn.Conv2d`, `nn.MultiheadAttention`).
   - Implement idempotent `remove()` method to clean up all hooks when fuzzing concludes.

---

#### Step 1: Run Pytest
```bash
source .venv/bin/activate
pip install torch
pytest -v tests/test_fuzzer.py
```

#### Step 2: Run Executable Verification Script
Save and run `scratch/verify_hook_memory.py`:
```python
import torch
import torch.nn as nn
from aegis.fuzzer import ActivationHookManager

class MultiLayerNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 16, 3, padding=1)
        self.relu = nn.ReLU()
        self.pool = nn.AdaptiveAvgPool2d((1, 1))
        self.fc = nn.Linear(16, 5)
    def forward(self, x):
        return self.fc(self.pool(self.relu(self.conv(x))).flatten(1))

model = MultiLayerNet()
hook_mgr = ActivationHookManager(model)

# Forward pass with test tensor
x = torch.randn(4, 3, 32, 32)
_ = model(x)

print(f'Captured {len(hook_mgr.current_activations)} layer activations:')
for name, stat in hook_mgr.current_activations.items():
    print(f'  Layer: {name:<10} L_inf: {stat.l_inf_norm:.4f} Mean: {stat.mean_activation:.4f}')

assert 'conv' in hook_mgr.current_activations
assert 'fc' in hook_mgr.current_activations

# Verify clean removal
hook_mgr.remove()
assert len(hook_mgr.hooks) == 0
assert len(hook_mgr.current_activations) == 0
print('TEST PASSED: Forward hook lifecycle verified with zero memory leaks!')
```
